src/Collection.php

```diff
@@ -810,6 +812,58 @@ public function listIndexes(array $options = [])
     }
 
     /**
+     * Executes a map-reduce aggregation on the collection.
+     *
+     * @see MapReduce::__construct() for supported options
+     * @see http://docs.mongodb.org/manual/reference/command/mapReduce/
+     * @param JavascriptInterface $map            Map function
+     * @param JavascriptInterface $reduce         Reduce function
+     * @param string|array|object $out            Output specification
+     * @param array               $options        Command options
+     * @return MapReduceResult
+     * @throws UnsupportedException if options are not supported by the selected server
+     * @throws InvalidArgumentException for parameter/option parsing errors
+     * @throws DriverRuntimeException for other driver errors (e.g. connection errors)
+     * @throws UnexpectedValueException if the command response was malformed
+     */
+    public function mapReduce(JavascriptInterface $map, JavascriptInterface $reduce, $out, array $options = [])
+    {
+        if ( ! isset($options['readPreference'])) {
+            $options['readPreference'] = $this->readPreference;
+        }
+
+        // Check if the out option is inline because we will want to coerce a primary read preference if not
+        if ( ! is_array($out) && ! is_object($out)) {
+            return false;
+        }
+
+        $out = (array) $out;
+
+        if (key($out) !== 'inline') {
+            $options['readPreference'] = new ReadPreference(ReadPreference::RP_PRIMARY);
+        }
+
+        $server = $this->manager->selectServer($options['readPreference']);
+
+        // Set options if not set already
```

Comment:
I would suggest copying the same comment from `aggregate()`, as the explanation for readConcern is the same (see: [SERVER-20214](https://jira.mongodb.org/browse/SERVER-20214)).

```
/* A "majority" read concern is not compatible with the $out stage, so
 * avoid providing the Collection's read concern if it would conflict.
 */
```

Also, a majority read concern is only a problem if `$out` is not inline (i.e. mapReduce will write results to an output collection). To that end, we probably want to track `$isOutInline` (or its inverse, `$hasOutputCollection`) and implement similar logic as `aggregate()` does with the `$hasOutStage` variable. For `aggregate()`, `$hasOutStage` means the command will write to a collection.